elixir/apps/domain/lib/domain/config/clients_upstream_dns.ex

```diff
@@ -0,0 +1,105 @@
+defmodule Domain.Config.ClientsUpstreamDNS do
+  @moduledoc """
+  Embedded Schema for Clients Upstream DNS
+  """
+  use Domain, :schema
+  import Domain.Validator
+  import Ecto.Changeset
+
+  @dns_types ~w[ip dns_over_tls dns_over_http]
+
+  @primary_key false
+  embedded_schema do
+    field :type, :string
+    field :address, :string
+  end
+
+  def changeset(dns_config \\ %__MODULE__{}, attrs) do
+    dns_config
+    |> cast(attrs, [:type, :address])
+    |> validate_required([:type, :address])
+    |> trim_change(:address)
+    |> validate_address()
+  end
+
+  defp validate_address(changeset) do
+    {_origin, type} = fetch_field(changeset, :type)
+
+    validate_change(changeset, :address, fn :address, address ->
+      case type do
+        "ip" ->
+          case Domain.Types.IPPort.cast(address) do
+            {:ok, _ip} ->
+              []
+
+            {:error, _reason} ->
+              [address: "must be a valid IP address"]
+          end
+
+        "dns_over_tls" ->
+          [address: "DNS over TLS is not supported yet"]
+
+        "dns_over_http" ->
+          [address: "DNS over HTTP is not supported yet"]
+
+        _other ->
+          [address: "Invalid Type"]
+      end
+    end)
+  end
+
+  # def normalize_dns_address(dns) do
+  #  case Domain.Types.IPPort.cast(dns.address) do
+  #    {:ok, ip} ->
+  #      port = ip.port || 53
+  #      %{ip | port: port} |> to_string()
+
+  #    {:error, _reason} ->
+  #      dns.address
+
+  #    :error ->
+  #      dns.address
+  #  end
+  # end
+
+  def normalize_dns_address(%__MODULE__{type: "ip", address: address}) do
+    case Domain.Types.IPPort.cast(address) do
+      {:ok, ip} ->
+        port = ip.port || 53
```

Comment:
Let's add `Domain.Types.IPPort.put_default_port/2`.

Reply:
Let's move `53` to a function called `default_ip_dns_port` (or something like that):
```
  def default_ip_dns_port, do: 53
```

Reply:
Added both of these suggestions.